File: tools/wasm-exception-jsc-diff.py
#!/usr/bin/env python3
"""Compare the public WebAssembly exception JS API with system JavaScriptCore."""

from __future__ import annotations

import difflib
import hashlib
import platform
import subprocess
import sys
import tempfile
from pathlib import Path

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = ROOT / "tests/wasm_exception_jsc_diff.c"


def run(command: list[str]) -> bytes:
    return subprocess.run(command, cwd=ROOT, check=True, stdout=subprocess.PIPE).stdout


def main() -> None:
    if len(sys.argv) != 2:
        raise SystemExit("usage: wasm-exception-jsc-diff.py <zig-js fixture>")
    if platform.system() != "Darwin":
        raise SystemExit("the WebAssembly exception JavaScriptCore differential requires macOS")

    with tempfile.TemporaryDirectory(prefix="zig-js-wasm-exception-jsc-") as directory:
        reference = Path(directory) / "wasm-exception-jsc"
        subprocess.run(
            [
                "xcrun", "--sdk", "macosx", "clang", str(FIXTURE),
                "-framework", "JavaScriptCore", "-o", str(reference),
            ],
            cwd=ROOT,
            check=True,
        )
        expected = run([str(reference)])
        actual = run([sys.argv[1]])

    if actual != expected:
        difference = difflib.unified_diff(
            expected.decode().splitlines(),
            actual.decode().splitlines(),
            fromfile="system JavaScriptCore",
            tofile="zig-js",
            lineterm="",
        )
        raise SystemExit("WebAssembly exception JSC differential mismatch:\n" + "\n".join(difference))
    digest = hashlib.sha256(actual).hexdigest()[:16]
    print(f"WebAssembly exception JSC differential: matched {len(actual.splitlines())} rows ({digest})")


if __name__ == "__main__":
    main()

File: tools/wasm-exception-jsc-diff.ts
/** Compare the public WebAssembly exception JS API with system JSC. */
import { checked, removeTemporaryDirectory, sha256File, temporaryDirectory, writeText, run } from "./lib/home";
declare const __dirname: string;
const ROOT = __dirname === "tools" ? "." : __dirname.slice(0, __dirname.lastIndexOf("/tools"));
const join = (left: string, right: string): string => `${left.replace(/\/$/, "")}/${right}`;
const args = process.argv.slice(2);
if (args.length !== 1) throw new Error("usage: wasm-exception-jsc-diff.ts <zig-js fixture>");
if (checked(["uname", "-s"], "detect platform").trim() !== "Darwin") throw new Error("the WebAssembly exception JavaScriptCore differential requires macOS");
const directory = temporaryDirectory("zig-js-wasm-exception-jsc");
try {
  const reference = join(directory, "wasm-exception-jsc");
  checked(["xcrun", "--sdk", "macosx", "clang", join(ROOT, "tests/wasm_exception_jsc_diff.c"), "-framework", "JavaScriptCore", "-o", reference], "compile system JSC WebAssembly exception fixture");
  const expected = checked([reference], `run ${reference}`), actual = checked([args[0]], `run ${args[0]}`);
  if (actual !== expected) {
    const expectedPath = join(directory, "system-jsc.txt"), actualPath = join(directory, "zig-js.txt");
    writeText(expectedPath, expected); writeText(actualPath, actual);
    const difference = run(["diff", "-u", expectedPath, actualPath]);
    throw new Error(`WebAssembly exception JSC differential mismatch:\n${difference.stdout || difference.stderr}`);
  }
  const output = join(directory, "actual.txt"); writeText(output, actual);
  const rows = actual ? actual.replace(/\n$/, "").split("\n").length : 0;
  console.log(`WebAssembly exception JSC differential: matched ${rows} rows (${sha256File(output).slice(0, 16)})`);
} finally { removeTemporaryDirectory(directory); }
